vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/TransactionImpl.java

```diff
@@ -58,21 +64,61 @@
     }
   }
 
-  @Override
-  protected <T> Promise<T> promise() {
-    return context.promise();
+  Future<Transaction> begin() {
+    PromiseInternal<Transaction> promise = context.promise(this::afterBegin);
+    ScheduledCommand<Transaction> b = doQuery(new TxCommand<>(TxCommand.Kind.BEGIN, this), promise);
+    doSchedule(b.cmd, b.handler);
+    return promise.future();
   }
 
-  @Override
-  protected <T> Promise<T> promise(Handler<AsyncResult<T>> handler) {
-    return context.promise(handler);
+  private <R> void doSchedule(CommandBase<R> cmd, Handler<AsyncResult<R>> handler) {
+    connection.schedule(cmd, context.promise(handler));
   }
 
-  private <R> void doSchedule(CommandBase<R> cmd, Handler<AsyncResult<R>> handler) {
-    conn.schedule(cmd, context.promise(handler));
+  private <R> void wrapAndSchedule(ScheduledCommand<R> scheduled) {
+    CommandBase<R> cmd = scheduled.cmd;
+    if (isComplete(cmd)) {
+      status = ST_COMPLETED;
+      doSchedule(cmd, ar -> {
+        if (ar.succeeded()) {
+          if (cmd == COMMIT) {
+            completionPromise.tryComplete();
+          } else {
+            completionPromise.tryFail(TransactionRollbackException.INSTANCE);
```

Comment:
we can let the `TransactionRollbackException` include the failure context here, like `new TransactionRollbackException(ar.cause)`.

Reply:
my bad, it's not possible here